SSA/Projects/DCE/DCE.lean

```diff
@@ -0,0 +1,492 @@
+  Γ.removeNth v.val
+
+/-- Witness that Γ' is Γ without v -/
+def Deleted {α : Ty} (Γ: Ctxt Ty) (v : Γ.Var α) (Γ' : Ctxt Ty) : Prop :=
+  Γ' = Γ.delete v
+
+/-- build a `Deleted` for a `(Γ.snoc α) → Γ`-/
+def Deleted.deleteSnoc (Γ : Ctxt Ty) (α : Ty) : Deleted (Γ.snoc α) (Ctxt.Var.last Γ α) Γ := rfl
+
+theorem List.removeNth_zero : List.removeNth (List.cons x xs) 0 = xs := rfl
+
+theorem List.removeNth_succ : List.removeNth (List.cons x xs) (.succ n) = x :: List.removeNth xs n := rfl
+
+/- removing from `xs ++ [x]` at index `(length xs)` equals `xs`. -/
+theorem List.removeNth_eq_len_snoc (hn : n = xs.length) : List.removeNth (xs ++ [x]) n = xs := by
+  induction n generalizing xs x
+  case zero =>
+    induction xs
+    case nil => simp
+    case cons x xs' IH => simp at hn
+  case succ n' IH' =>
+    induction xs
+    case nil => simp at hn
+    case cons x xs' IH =>
+      simp[removeNth_succ]
+      apply IH'
+      simp at hn
+      exact hn
+
+/-- removing at any index `≥ xs.length` does not change the list. -/
+theorem List.removeNth_of_length_le (hn : xs.length ≤ n) : List.removeNth xs n = xs := by
+  induction n generalizing xs
+  case zero =>
+    induction xs
+    case nil => simp
+    case cons x xs' IH => simp at hn
+  case succ n' IH' =>
+    induction xs
+    case nil => simp[removeNth]
+    case cons x xs' IH =>
+      simp[removeNth_succ]
+      apply IH'
+      simp at hn
+      linarith
+
+/- removing at index `n` does not change indices `k < n` -/
+theorem List.get?_removeNth_of_lt (hk: k < n) : List.get? (List.removeNth xs n) k = List.get? xs k := by
+  by_cases N_LEN:(xs.length ≤ n)
+  case pos => simp[removeNth_gt_len N_LEN]
+  case neg =>
+    simp at N_LEN
+    induction xs generalizing n k
+    case nil => simp at N_LEN
+    case cons hd tl IHxs =>
+    cases n
+    case zero => simp at hk
+    case succ n' =>
+      simp[List.removeNth_succ]
+      cases k
+      case zero => simp
+      case succ k' =>
+        simp[List.get?]
+        apply IHxs
+        linarith
+        simp at N_LEN; linarith
+
+
+/-- Removing index `n` shifts entires of `k ≥ n` by 1. -/
+theorem List.get?_removeNth_of_le {xs : List α} {n : Nat} {k : Nat} (hk: n ≤ k) :
+  (xs.removeNth n).get? k = xs.get? (k + 1) := by
+  induction xs generalizing n k
+  case nil => simp[removeNth, List.get]
+  case cons hd tl IHxs =>
+    simp[List.get];
+    cases k
+    case zero =>
+      simp at hk
+      subst hk
+      simp[removeNth]
+    case succ k' =>
+      cases n
+      case zero =>
+        simp[List.removeNth_succ]
+      case succ n' =>
+        simp[List.removeNth_succ]
+        apply IHxs
+        linarith
+
+/-- Lean bug: linarith fails to find a contradiction -/
+theorem linarith_failure (x y : Nat) (H1 : ¬ (x = y)) (H2 : ¬ (x < y)) : x > y :=
+  by
+    try linarith
+    sorry
+
+/-- Given  `Γ' := Γ /delv`, transport a variable from `Γ'` to `Γ`. -/
+def Deleted.pullback_var (DEL : Deleted Γ delv Γ') (v : Γ'.Var β) : Γ.Var β :=
```

Comment:
Why have this go through `Deleted`, instead of direclty writing `(Γ.delete v).Var β -> Γ.Var β`?

Reply:
Because `Deleted` acts as a hook for keeping these theorems.